Added a minimal working/functional GTK UI for analyzing fuzzing stats

DMaroo opened the pull request:
(This PR resulted from the first idea mentioned in [ideas.md](https://github.com/AFLplusplus/AFLplusplus/blob/stable/docs/ideas.md). Please let me know if I misunderstood the scope or the purpose of that idea.)

Added a very simple and minimal UI (created using GTK) for analyzing the graphs plotted using `afl-plot`. This allows to zoom in and out of the graphs, scroll through them, and also shows the cursor coordinates. I have recorded the demo to give some more idea about it.

https://user-images.githubusercontent.com/72350048/128217651-46d0e90d-d2a6-4ee4-9071-ee22601eeed5.mp4

This pull request will (hopefully) be one of many pull requests that I intend to make (if the maintainers permit) to improve upon this basic UI. Some rough TODOs could be as follows:

 - Adding buttons and scrollbars
 - Improving the *"look-and-feel"* of the UI
 - Providing more support as mentioned in [ideas.md](https://github.com/AFLplusplus/AFLplusplus/blob/stable/docs/ideas.md)

DMaroo committed:
Wrote a simple and minimal UI in GTK, with GNUplot graphs embedded in it

TODO: Improve upon the extremely minimal (yet functional) UI

DMaroo committed:
Updated the afl-plot shell script to use GTK UI

TODO: Currently the script is very huge, and needs to be made more
concise and clean. In it's current state, it is hard to maintain.

vanhauser-thc commented:
hi,
please move the source code file to a new folder like utils/plot_ui - that will also make solving your Makefile issue much easier as you can provide your own.
for afl-plot, add code to check if afl-plot-ui is present and use it if so, and do the html stuff otherwise. plus a command line option to enforce the html no ui .
thank you

DMaroo committed:
Added usage info and command line help flag support

TODO: Prevent standalone execution

DMaroo committed:
Added Makefile for building and installing afl-plot-ui

Currently, it offers make and make clean functionality only

DMaroo committed:
Added README to describe the working of afl-plot-ui

DMaroo committed:
Added support for graphical command line flag, improved the general organization of script

Script works as "intended"

DMaroo committed:
Added build and install instructions for afl-plot-ui

I am implicitly assuming that the user already has GK and pkg-config
installed, as requested in the build instructions in the main README

DMaroo committed:
Added afl-plot-ui executable in .gitignore list

DMaroo committed:
Updated README to contain description regarding plot_ui

DMaroo committed:
Added additional dependencies which need to be installed

DMaroo commented:
@vanhauser-thc I have completed this pull request in my understanding. Please suggest any necessary changes. Thank you.

vanhauser-thc commented on afl-plot:
```diff
@@ -22,16 +22,21 @@ get_abs_path() {
 
   cat 1>&2 <<_EOF_
-$0 afl_state_dir graph_output_dir
+$0 afl_state_dir graph_output_dir [ -g | --graphical ]
```
switches like `-g` are first, not last.
just do 
```
if [ "$1" = "-g" ]; then
  ...
  shift # shifts $2 and $3 to $1 and $2
fi
```

vanhauser-thc commented on GNUmakefile:
```diff
@@ -601,6 +602,7 @@ distrib: all
 	$(MAKE) -C utils/afl_network_proxy
 	$(MAKE) -C utils/socket_fuzzing
 	$(MAKE) -C utils/argv_fuzzing
+	$(MAKE) -C utils/plot_ui
```
a) MUST have a `-` in front: `-$(MAKE)` is the build can fail if the packages are not installed and the the whole afl++ build fails.
b) I am not sure it should always be built. I would propose to first put it in for people having to manual build it if they want to for the dev branch and then asking on the fuzzing discord channel who useful people find it, and if the majority find it useful then another PR that always builds it.

DMaroo replied:
Thank you for informing about point *a)*.

I am sorry, but I am unable to understand what you're trying to say in point *b)*. Should I create a new make target in which `afl-plot-ui` will be built? I have opened the PR for the `AFLplusplus:dev` branch, so I don't suppose this will be merged in the stable branch. I am not able to see what exactly I should be changing in the Makefile.

vanhauser-thc replied:
I am basically saying, move this from
```
$(MAKE) -C utils/plot_ui
```
to
```
#-$(MAKE) -C utils/plot_ui
```

DMaroo replied:
Okay, got it. I have now commented out all the lines where I was building `utils/plot_ui`.

vanhauser-thc commented on afl-plot:
```diff
@@ -137,94 +124,105 @@ set key outside
-     '' using 1:8 with lines title ' uniq crashes' linecolor rgb '#c00080' linewidth 3, \\
-     '' using 1:9 with lines title 'uniq hangs' linecolor rgb '#c000f0' linewidth 3, \\
-     '' using 1:10 with lines title 'levels' linecolor rgb '#0090ff' linewidth 3
+if [ "$#" = "3" ] && ( [ "$3" = "-g" ] || [ "$3" = "--graphical" ] ); then
```
IMHO it is fine for the `else` of this print to give an informational message that "-g" would give a graphical UI if afl-plot-ui is built and installed. otherwise people wont know about it.

DMaroo committed:
Updated the CI configuration to install the new dependencies introduced

DMaroo committed:
Changed parameter order, reduced script execution time

vanhauser-thc commented on afl-plot:
```diff
@@ -285,6 +291,8 @@ _EOF_
 
 ) | gnuplot
 
+echo "[?] You can also use -g flag to view the plots in an GUI window, and interact with the plots (if you have build afl-plot-ui). Run \"afl-plot\" to know more."
```
typo, should be "if you have built", also switch "Run \"afl-plot\"" to "Run \"afl-plot -h\"". thanks

DMaroo committed:
Commented out all the builds of afl-plot-ui

Also add the "-" qualifier to denote a possibly unsuccessful build

DMaroo committed:
Fixed a typo

DMaroo committed:
Removed the optional build dependencies for afl-plot-ui

DMaroo committed:
Added separate build and install instructions for afl-plot-ui

DMaroo commented:
@vanhauser-thc I have also updated `README.md` with separate instructions for building afl-plot-ui. Please suggest if any further changes are needed. Thank you.

vanhauser-thc commented:
LGTM, thanks

vanhauser-thc commented:
some feedback:

please add the required apt-get command to the README of the tool.

The UI is nice, is it possible to start it maximized? when I started it, it was a super tiny window.

=> https://discord.com/channels/736989425770168422/736989426298912813/873508410853838868

DMaroo commented:
Thank you for the feedback. I have incorporated both the suggestions and opened a PR (#1055).

Also, if you don't mind, could you please open an issue containing of a checklist of all the improvements or developments to be done for the current minimal UI. The ideas.md is too terse for me to understand concrete objectives to work on. A checklist in form an issue would be very helpful.

Secondly, according to the current setup of things, I don't think it is possible to add zoom in/out buttons or scrollbars in the current UI, since zooming in/out and scrolling requires controlling GNUplot. In the current setup, the UI is completely separated from GNUplot, and cannot interact with GNUplot. We could use any message passing library to communicate between the UI and the shell script (which controls GNUplot), but that seems unnecessarily too complex to me, as of now. It will be fine to add any more windows, but adding more interactivity (other than zooming in/out, scrolling and cursor positioning) could be a problem given the current setup. I tried to incorporate the GUI in the least intrusive way possible, which is why it will be hard to integrate it to provide more interactivity.
